### ISSUE TITLE
io_uring_register_buffers and io_uring_register_buf_ring

### ISSUE THREAD

**Issue by lorislibralato** [axboe/liburing#1025]
Does registering the buffers used in the `buf_ring` give performance improvements with multishot operations?

**Comment by isilence** [comment]
If we compare with the legacy provided buffers, i.e. `IORING_OP_PROVIDE_BUFFERS`, then yes, `buf_ring` is more performant, not needing to queue extra requests is a huge win.

But note that to use multishot receives in the first place you need one of those -- multishots potentially run indefinitely and you need a way to feed it buffers.

**Comment by lorislibralato** [comment]
I didn't explained well, If i use io_uring_register_buffer on a buffer added to a buf_ring does It optimize the buffer operations?

**Comment by isilence** [comment]
> I didn't explained well, If i use io_uring_register_buffer on a buffer added to a buf_ring does It optimize the buffer operations?

No, it does not. The kernel doesn't keep track of that association.

If one day there would be providing  registered buffers of some sort it should be explicitly reflected in the user API.  And nobody cared to implement it because it wouldn't help anyway for the same reason why registered buffers are useless for recv / recvmsg -- there is a memcpy inside the net stack.

A side note, there is a zerocopy recv in development, it requires registered buffers, and long merged send zc also benefits from registered buffers, but

**Comment by lorislibralato** [comment]
Ok thanks!
